src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioRazorParser.cs

```diff
@@ -0,0 +1,164 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.VisualStudio.Text;
+using ITextBuffer = Microsoft.VisualStudio.Text.ITextBuffer;
+using Timer = System.Threading.Timer;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    public class VisualStudioRazorParser : ForegroundThreadAffinitizedObject
+    {
+        // Internal for testing.
+        internal readonly ITextBuffer _textBuffer;
+        internal int IdleDelay = 3000;
+
+        private readonly IRazorCompletionBroker _completionBroker;
+        private readonly BackgroundParser _parser;
+        private Timer _idleTimer;
+        private RazorSyntaxTreePartialParser _partialParser;
+
+        public VisualStudioRazorParser(ITextBuffer buffer, RazorTemplateEngine templateEngine, string filePath, IRazorCompletionBroker completionBroker)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (templateEngine == null)
+            {
+                throw new ArgumentNullException(nameof(templateEngine));
+            }
+
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(
+                    AspNetCore.Razor.Language.Resources.ArgumentCannotBeNullOrEmpty,
+                    nameof(filePath));
+            }
+
+            if (completionBroker == null)
+            {
+                throw new ArgumentNullException(nameof(completionBroker));
+            }
+
+            TemplateEngine = templateEngine;
+            FilePath = filePath;
+            _textBuffer = buffer;
+            _completionBroker = completionBroker;
+            _textBuffer.Changed += TextBuffer_OnChanged;
+            _parser = new BackgroundParser(templateEngine, filePath);
+            _parser.ResultsReady += OnResultsReady;
+
+            _parser.Start();
+        }
+
+        public event EventHandler<DocumentStructureChangedEventArgs> DocumentStructureChanged;
+
+        public RazorTemplateEngine TemplateEngine { get; }
+
+        public string FilePath { get; }
+
+        public RazorCodeDocument CodeDocument { get; private set; }
```

Comment:
If I want to access the 'current' state, how do I know what snapshot it's related to?